hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/client/TestHoodieSparkIndex.java

```diff
@@ -0,0 +1,154 @@
+
+    //validate rest.
+    HoodieTableMetaClient newMetaClient = HoodieTableMetaClient.builder().setBasePath(metaClient.getBasePath()).setConf(metaClient.getStorageConf())
+        .build();
+    HoodieIndexDefinition indexDefinition = getHoodieIndexDefn(indexNamePrefix + "_0", PARTITION_NAME_SECONDARY_INDEX, fieldNamePrefix + "_0");
+    readAndValidateIndexDefn(indexDefinition, newMetaClient);
+    indexDefinition = getHoodieIndexDefn(indexNamePrefix + "_3", PARTITION_NAME_EXPRESSION_INDEX, fieldNamePrefix + "_3");
+    readAndValidateIndexDefn(indexDefinition, newMetaClient);
+    indexDefinition = getHoodieIndexDefn(indexNamePrefix + "_4", PARTITION_NAME_SECONDARY_INDEX, fieldNamePrefix + "_4");
+    readAndValidateIndexDefn(indexDefinition, newMetaClient);
+    readAndValidateIndexDefn(colStatsIndexDefinition);
+
+    // update col stats w/ new set of cols
+    List<String> colsToIndex = IntStream.range(0, 10).mapToObj(number -> fieldNamePrefix + "_" + number).collect(Collectors.toList());
+    colStatsIndexDefinition = getHoodieIndexDefn(PARTITION_NAME_COLUMN_STATS,
+        RANGE_TYPE, RANGE_TYPE, colsToIndex, Collections.EMPTY_MAP);
+    sparkIndexClient.register(metaClient, colStatsIndexDefinition);
+    readAndValidateIndexDefn(colStatsIndexDefinition);
+
+    // drop col stats
+    metaClient.deleteIndexDefinition(colStatsIndexDefinition.getIndexName());
+    readAndValidateIndexDefnNotPresent(colStatsIndexDefinition, HoodieTableMetaClient.builder().setBasePath(metaClient.getBasePath()).setConf(metaClient.getStorageConf())
+        .build());
+  }
+
+  private void readAndValidateIndexDefn(HoodieIndexDefinition expectedIndexDefn) {
+    HoodieTableMetaClient newMetaClient = HoodieTableMetaClient.builder().setBasePath(metaClient.getBasePath()).setConf(metaClient.getStorageConf())
+        .build();
+    readAndValidateIndexDefn(expectedIndexDefn, newMetaClient);
+  }
+
+  private void readAndValidateIndexDefn(HoodieIndexDefinition expectedIndexDefn, HoodieTableMetaClient metaClient) {
+    assertTrue(metaClient.getIndexMetadata().isPresent());
+    assertTrue(!metaClient.getIndexMetadata().get().getIndexDefinitions().isEmpty());
+    assertTrue(metaClient.getIndexMetadata().get().getIndexDefinitions().containsKey(expectedIndexDefn.getIndexName()));
+    assertEquals(expectedIndexDefn, metaClient.getIndexMetadata().get().getIndexDefinitions().get(expectedIndexDefn.getIndexName()));
+  }
+
+  private void readAndValidateIndexDefnNotPresent(HoodieIndexDefinition expectedIndexDefn, HoodieTableMetaClient metaClient) {
+    assertTrue(metaClient.getIndexMetadata().isPresent());
+    assertTrue(!metaClient.getIndexMetadata().get().getIndexDefinitions().isEmpty());
+    assertTrue(!metaClient.getIndexMetadata().get().getIndexDefinitions().containsKey(expectedIndexDefn.getIndexName()));
+  }
+
+  private HoodieIndexDefinition getHoodieIndexDefn(String indexName, String indexType, String sourceField) {
```

Comment:
also, for other methods, let's rename `*HoodieIndexDefn*` to `*IndexDefinition*`